### PR TITLE
Convert heaviside step function component to expression

### PR DESCRIPTION
**Pull request by jlaehne**
### Description of the change
Conversion of the heaviside step function component to an expression-based definition. As expression allows for only one `where`, this version follows the alternative definition without intermediate value at `x=n`. 

### Progress of the PR
- [X] Change implemented,
- [X] update docstring,
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
>>> import hyperspy.api as hs
>>> h1 = hs.model.components1D.HeavisideStep(A=3,n=2)
>>> x = np.arange(0,5,step=0.01)
>>> s = hs.signals.Signal1D(h1.function(x))
>>> s.axes_manager.signal_axes[0].axis = x
>>> s.plot()
>>> h1.print_current_values()
>>> # Your new feature...
```
Note that this example can be useful to update the user guide.

### CODE REVIEW

**Comment by jlaehne** [comment]
As noted by @francisco-dlp in #2251, an alternative for the definition would be using numpy's `heaviside` function. However, this function does not allow a change of the zero point, so one would have to pamper with the x-axis to achieve a step at a different position than zero.

Note that `np.heaviside` does not necessarily include the intermediate value at the step either:
`np.heaviside(np.arange(-1,1.5,0.5), 0.5)` versus `np.heaviside(np.arange(-1,1.2,0.2), 0.5)`.

**Comment by jlaehne** [comment]
At the moment, the problem with `heaviside` defined as expression is that  `expression.py` cannot automatically calculate the gradient of the step function. Any ideas on that?

File: doc/dev_guide/git.rst [hunk 1]
```diff
@@ -5,60 +5,60 @@
...
 ====================
 
-For developing the code, the home of HyperSpy is on 
-`GitHub <https://github.com>`_, and you'll see that
-a lot of this guide boils down to properly use that platform. So, visit the
+For developing the code the home of HyperSpy is on github and you'll see that
+a lot of this guide boils down to using that platform well. So visit the
 following link and poke around the code, issues, and pull requests: `HyperSpy
-on GitHub <https://github.com/hyperspy/hyperspy>`_.
+on Github <https://github.com/hyperspy/hyperspy>`_.
```
Comment (dnjohnstone) [inline doc/dev_guide/git.rst]: You've used Github here at github elsewhere - it seems you were looking to standardise, so standardise.

File: doc/dev_guide/git.rst [hunk 2]
```diff
@@ -5,60 +5,60 @@
 Using Git and GitHub
 ====================
 
-For developing the code, the home of HyperSpy is on 
-`GitHub <https://github.com>`_, and you'll see that
-a lot of this guide boils down to properly use that platform. So, visit the
+For developing the code the home of HyperSpy is on github and you'll see that
```
Comment (dnjohnstone) [inline doc/dev_guide/git.rst]: This is still odd wording... how about "The home of HyperSpy code development is on github..."

File: doc/dev_guide/git.rst [hunk 3]
```diff
@@ -5,60 +5,60 @@
...
+By now you'll have had a look around GitHub - but why's it so important?
 
-Well, GitHub is the public forum in which we manage and discuss development of
-the code. More importantly, it enables every developer to use `Git`, which is
-an open source "version control" system. By version control, we mean that you
-can separate out your contribution to the code into many versions (called
-branches) and switch between them easily. Later, you can choose which version
-you want to have integrated into HyperSpy. You can learn all about Git at
-`git-scm <http://www.git-scm.com/about>`_!
+Well GitHub is the public forum in which we manage and discuss development of
```
Comment (dnjohnstone) [inline doc/dev_guide/git.rst]: Why not just delete the "Well", and again with the different forms of GitHub capitalization.

File: doc/dev_guide/git.rst [hunk 4]
```diff
@@ -5,60 +5,60 @@
...
+the code. More importantly, it enables every developer to utilise Git which is
```
Comment (dnjohnstone) [inline doc/dev_guide/git.rst]: I think this actually should be "use" and here you've used "Git" where elsewhere you've changed it to "git". There should also be a comma before a "which" here.

File: doc/dev_guide/git.rst [hunk 5]
```diff
@@ -5,60 +5,60 @@
...
+desktop. By version control we mean that you can separate out your contribution
+to the code into many versions (called branches) and switch between them
+easily. Later you can choose which version you want to have integrated into
+HyperSpy.
 
-It is very important to separate your contributions so
-that each branch is a small advancement on the "master" code or on another
-branch. In the end, each branch will have to be checked and reviewed by
-someone else before it can be included -- so if it is too big, you will be
+You can learn all about Git `here <http://www.git-scm.com/about>`_!
```
Comment (dnjohnstone) [inline doc/dev_guide/git.rst]: Are you using Git or git?

File: doc/dev_guide/git.rst [hunk 6]
```diff
@@ -5,60 +5,60 @@
...
+someone else before it can be included - so if it's too big, you will be
 asked to split it up!
 
 For personal use, before integrating things into the main HyperSpy code, you
-can merge some together for your personal use. However, make sure each new
-feature has it's own branch that is contributed through a separate pull
-request!
+may want to use a few branches together. You can do that but just make sure
+each new thing has it's own branch! You can merge some together for your
+personal use.
```
Comment (dnjohnstone) [inline doc/dev_guide/git.rst]: Isn't this final sentence a bit repetitive?

File: doc/dev_guide/index.rst [hunk 7]
```diff
@@ -10,7 +10,6 @@ HyperSpy Developer Guide
     :maxdepth: 2
 
     intro.rst
-    git.rst
```
Comment (dnjohnstone) [inline doc/dev_guide/index.rst]: Why is that being removed?

File: doc/dev_guide/intro.rst [hunk 8]
```diff
@@ -17,22 +17,23 @@ Getting started
...
 HyperSpy works and to build a broad overview of the code as it stands is to
 use it -- so what are you waiting for? `Install HyperSpy
 <http://hyperspy.org/hyperspy-doc/current/user_guide/install.html>`_.
 
-The `HyperSpy User-Guide <http://www.hyperspy.org/hyperspy-doc/current/index
-.html>`_ also provides a good overview of all the parts of the code that
+The user-guide also provides a good overview of all the parts of the code that
 are currently implemented as well as much information about how everything
-works -- so read it well.
+works -- so read it well:
```
Comment (dnjohnstone) [inline doc/dev_guide/intro.rst]: -- to - if you're doing that elsewhere

File: doc/dev_guide/intro.rst [hunk 9]
```diff
@@ -17,22 +17,23 @@ Getting started
...
+`HyperSpy User-Guide <http://www.hyperspy.org/hyperspy-doc/current/index
+.html>`_.
 
 
 2. Got a problem? -- ask!
```
Comment (dnjohnstone) [inline doc/dev_guide/intro.rst]: -- to -

File: doc/dev_guide/intro.rst [hunk 10]
```diff
@@ -50,17 +51,17 @@ they're logged until dealt with. This is also a good place to make a proposal
 for some new feature or tool that you want to work on.
 
 
-3. Contribute -- yes you can!
------------------------------
+3. Contribute—yes you can!
```
Comment (dnjohnstone) [inline doc/dev_guide/intro.rst]: spaces around the -

File: doc/dev_guide/intro.rst [hunk 11]
```diff
@@ -78,30 +79,30 @@ Indeed, there are many ways to contribute:
...
 requests on the `GitHub Page <https://github.com/hyperspy/hyperspy/>`_.
 You'll find that there are many known areas for development in the issues
-and a number of pull-requests are partially finished projects just sitting 
-there waiting for a keen new contributor to come and learn by finishing.
+and a number of pull-requests are part finished projects just sitting there
+waiting for a keen new contributor to come and learn by finishing.
 
 The documentation (let it be the docstrings,
-guides or the website) is always in need of some care. Besides,
+guides or the website) is always in need of some care. Beside,
```
Comment (dnjohnstone) [inline doc/dev_guide/intro.rst]: Just delete the "Beside(s),"

File: doc/dev_guide/lazy_computations.rst [hunk 12]
```diff
@@ -12,11 +12,11 @@ the two arrays are indeed almost identical, the most important differences are
 (``da`` being ``dask.array.Array`` in the examples):
 
  - **Dask arrays are immutable**: ``da[3] = 2`` does not work. ``da += 2``
-   does, but it's actually a new object -- you might as well use ``da = da + 2``
+   does, but it's actually a new object -- might as well use ``da = da + 2``
```
Comment (dnjohnstone) [inline doc/dev_guide/lazy_computations.rst]: -- to -

File: doc/dev_guide/testing.rst [hunk 13]
```diff
@@ -58,13 +57,13 @@ Useful hints on testing:
 * The ``hyperspy.misc.test_utils.py`` contains a few useful functions for
   testing.
 * ``@pytest.mark.parametrize()`` is a very convenient decorator to test several
-  parameters of the same function without having to write too much repetitive
+  parameters of the same function without having to write to much repetitive
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: this should be "too"

File: doc/dev_guide/testing.rst [hunk 14]
```diff
@@ -58,13 +57,13 @@ Useful hints on testing:
...
   code, which is often error-prone. See `pytest documentation for more details
   <http://doc.pytest.org/en/latest/parametrize.html>`_.
-* It is good to check that the tests do not use too much memory after
+* It is good to check that the tests does not use too much of memory after
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: This is wrong and it was correct before. What a test does affects what the tests do.

File: doc/dev_guide/testing.rst [hunk 15]
```diff
@@ -75,10 +74,10 @@ Useful hints on testing:
 * Once, you have pushed your PR to the official HyperSpy repository, it can be
   useful to check the coverage of your tests using the coveralls.io check of
   your PR. There should be a link to it at the bottom of your PR on the github
-  PR page. This service can help you to find out how well your code is being tested
+  PR page. This service can help you to find how well your code is being tested
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: find --> determine

File: doc/dev_guide/testing.rst [hunk 16]
```diff
@@ -75,10 +74,10 @@ Useful hints on testing:
...
   and exactly which part is not currently tested.
 * `pytest-sugar <https://pypi.python.org/pypi/pytest-sugar>`_ can be installed
-  to have a nicer look and feel of pytest in the console (encoding issues have
+  to have a nicer look and feel of pytest in the console (encoding issue have
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: This was correct before.

File: doc/dev_guide/testing.rst [hunk 17]
```diff
@@ -89,28 +88,28 @@ Plot testing
 Plotting is tested using the ``@pytest.mark.mpl_image_compare`` decorator of
 the `pytest mpl plugin <https://pypi.python.org/pypi/pytest-mpl>`_.  This
 decorator uses reference images to compare with the generated output during the
-tests. The reference images are located in the folder defined by the argument
+tests. The references images are located in the folder defined by the argument
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: This is wrong and was correct before.

File: doc/dev_guide/testing.rst [hunk 18]
```diff
@@ -89,28 +88,28 @@ Plot testing
...
 ``baseline_dir`` of the ``@pytest.mark.mpl_image_compare`` decorator.
 
-To run plot tests, you simply need to add the option ``--mpl``:
+To run plotting tests, you simply need to add the option ``--mpl``:
 ::
 
     pytest --mpl
 
-If you don't use ``--mpl``, the code of the tests will be executed, but the
+If you don't use the ``--mpl``, the code of the tests will be executed but the
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: I don't think you need the "the" but getting rid of the comma is good.

File: doc/dev_guide/testing.rst [hunk 19]
```diff
@@ -89,28 +88,28 @@ Plot testing
...
 images will not be compared to the references images.
 
 If you need to add or change some plots, follow the workflow below:
 
-    1. Write the tests using appropriate decorators such as
+    1. Write the tests using appropriate decorator such as
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: This was correct before

File: doc/dev_guide/testing.rst [hunk 20]
```diff
@@ -89,28 +88,28 @@ Plot testing
...
        ``@pytest.mark.mpl_image_compare``.
-    2. If you need to generate a new reference image in the folder
+    2. If you need to generate new reference image in the folder
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: If you're going to take the "a" away you need to make it "images"

File: doc/dev_guide/testing.rst [hunk 21]
```diff
@@ -89,28 +88,28 @@ Plot testing
...
        ``plot_test_dir``, for example, run: ``pytest
        --mpl-generate-path=plot_test_dir``
     3. Run again the tests and this time they should pass.
     4. Use ``git add`` to put the new file in the git repository.
 
-When the plot tests are failling, it is possible to download the figure 
+When the plotting tests are failling, it is possible to download the figure
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: Should be "failing" here

File: doc/dev_guide/testing.rst [hunk 22]
```diff
@@ -121,14 +120,14 @@ matplotlib generates the figure can make the tests fail.
 In travis and appveyor, the matplotlib backend is set to ``agg`` by setting
 the ``MPLBACKEND`` environment variable to ``agg``. At the first import of
 ``matplotlib.pyplot``, matplotlib will look at the ``MPLBACKEND`` environment
-variable and accordingly set the backend.
+variable and set accordingly the backend.
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: Should be "set the backend accordingly"

File: doc/dev_guide/testing.rst [hunk 23]
```diff
@@ -121,14 +120,14 @@ matplotlib generates the figure can make the tests fail.
...
 
 See `pytest-mpl <https://pypi.python.org/pypi/pytest-mpl>`_ for more details.
 
 Exporting pytest results as HTML
 ^^^^^^^^^^^^
 With ``pytest-html`` it is possible to export the results of running pytest 
-for easier viewing. It can be installed by conda:
+for easier viewing. I can be installed by conda:
```
Comment (dnjohnstone) [inline doc/dev_guide/testing.rst]: "It" made more sense

File: doc/dev_guide/writing_docs.rst [hunk 24]
```diff
@@ -7,8 +7,8 @@ Writing documentation
 
 Documentation comes in two parts: docstrings and user-guide documentation.
 
-Docstrings -- Are written at the start of a function and give essential information
-about how it should be used, such as, which arguments can be passed to it and
+Docstrings -- written at the start of a function and give essential information
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_docs.rst]: -- to -

File: doc/dev_guide/writing_extensions.rst [hunk 25]
```diff
@@ -23,30 +23,30 @@ HyperSpy. Because of HyperSpy's structure, we anticipate that most packages
 registering HyperSpy extensions will provide support for specific sorts of
 data.
 
-Models can also be provided by external packages, but don't need to
+Models can be provided by external packages too but don't need to
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: just delete the "too"

File: doc/dev_guide/writing_extensions.rst [hunk 26]
```diff
@@ -23,30 +23,30 @@ HyperSpy. Because of HyperSpy's structure, we anticipate that most packages
...
 It is good practice to add all packages that extend HyperSpy
 `to the list of known extensions
-<https://github.com/hyperspy/hyperspy-extensions-list>`_ regardless of their
-maturity level. In this way, we can avoid duplication of efforts and issues
+<https://github.com/hyperspy/hyperspy-extensions-list>`_ regardless their
+maturity level. In this way we can avoid duplication of efforts and issues
 arising from naming conflicts.
 
-At this point, it is worth noting that HyperSpy's main strength is its amazing
+At this point it is worth noting that HyperSpy's main strength is its amazing
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: Just delete the "At this point"

File: doc/dev_guide/writing_extensions.rst [hunk 27]
```diff
@@ -23,30 +23,30 @@ HyperSpy. Because of HyperSpy's structure, we anticipate that most packages
...
 community of users and developers. We trust that the developers of packages
 that extend HyperSpy will play by the same rules that have made the Python
 scientific ecosystem successful. In particular, avoiding duplication of
 efforts and being good community players by contributing code to the best
-matching project are essential for the sustainability of our open software
+matching project are essential for the sustainability of our software
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: I like the emphasis of "open" but not a big deal

File: doc/dev_guide/writing_extensions.rst [hunk 28]
```diff
@@ -59,15 +59,15 @@ In order to register HyperSpy extensions, you need to:
...
+For a full example on how to create a package that extends HyperSpy see
 `the HyperSpy Sample Extension package
 <https://github.com/hyperspy/hyperspy_sample_extension>`_.
 
 
 Creating new HyperSpy BaseSignal subclasses
 -------------------------------------------
 
-When and where to create a new ``BaseSignal`` subclass
+When and where create a new ``BaseSignal`` subclass
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: better with the "to"

File: doc/dev_guide/writing_extensions.rst [hunk 29]
```diff
@@ -90,14 +90,14 @@ for your function, you should consider creating your own.
 
        graph TD
 
-         A(New function needed!)
+         A(New function needed)
          B{Is it useful for data of any type and dimensions?}
          C(Contribute it to BaseSignal)
-         D{Does a SignalxD for the required dimension exist in HyperSpy?}
+         D{Does an SignalxD for the required dimension exist in HyperSpy?}
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: should be "a"

File: doc/dev_guide/writing_extensions.rst [hunk 30]
```diff
@@ -90,14 +90,14 @@ for your function, you should consider creating your own.
...
          E[Contribute new SignalxD to HyperSpy]
-         F{Is the function useful for a specific type of data only?}
+         F{Is the function useful only for some sort of data?}
          G(Contribute it to SignalxD)
-         H{Does a signal for that sort of data exists?}
+         H{Does an signal for that sort of data exists?}
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: should be "a"

File: doc/dev_guide/writing_extensions.rst [hunk 31]
```diff
@@ -115,14 +115,14 @@ for your function, you should consider creating your own.
 .. graphviz::
 
     digraph G {
-         A [label="New function needed!"]
+         A [label="New function needed"]
          B [label="Is it useful for data of any type and dimensions?",shape="diamond"]
          C [label="Contribute it to BaseSignal"]
-         D [label="Does a SignalxD for the required dimension exist in HyperSpy?",shape="diamond"]
+         D [label="Does an SignalxD for the required dimension exist in HyperSpy",shape="diamond"]
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: a

File: doc/dev_guide/writing_extensions.rst [hunk 32]
```diff
@@ -115,14 +115,14 @@ for your function, you should consider creating your own.
...
          E [label="Contribute new SignalxD to HyperSpy"]
-         F [label="Is the function useful for a specific type of data only?",shape="diamond"]
+         F [label="Is the function useful only for some sort of data?",shape="diamond"]
          G [label="Contribute it to SignalxD"]
-         H [label="Does a signal for that sort of data exist?",shape="diamond"]
+         H [label="Does an signal for that sort of data exists?",shape="diamond"]
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: this was correct before

File: doc/dev_guide/writing_extensions.rst [hunk 33]
```diff
@@ -187,8 +187,8 @@ acronyms for ``signal_type_aliases``.
 Creating new HyperSpy model components
 --------------------------------------
 
-When and where to create a new component
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When and where create a new component
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: better with the "to"

File: doc/dev_guide/writing_extensions.rst [hunk 34]
```diff
@@ -172,7 +172,7 @@ match for each sort of data.
 .. warning::
     HyperSpy assumes that only one signal
     subclass exists for a particular ``signal_type``. It is up to external
-    package developers to avoid ``signal_type`` clashes, typically by collaborating
+    packages developers to avoid ``signal_type`` clashes, typically by collaborating
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: this was correct before

File: doc/dev_guide/writing_extensions.rst [hunk 35]
```diff
@@ -201,8 +201,8 @@ HyperSpy itself (if they are of common interest) or to specialized external
 packages extending HyperSpy.
 
 The flowchart below can help you decide when and where to add
-a new hyperspy model :py:class:`hyperspy.component.Component`
-for your function, should you consider creating your own.
+a new hyperspy model :py:class:`hyperspy.component.Component`.
+for your function you should consider creating your own.
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: ok to get rid of the comma but the word order was better before

File: doc/dev_guide/writing_extensions.rst [hunk 36]
```diff
@@ -331,27 +329,27 @@ Registering toolkeys
...
 
 .. code-block:: yaml
 
     GUI:
       # In order to assign a widget to a function, that function must declare
       # a `toolkey`. The `toolkeys` list contains a list of all the toolkeys
-      # provided by extensions. In order to avoid name clashes, by convention,
-      # toolkeys must start with the name of the package that provides them.
+      # provided by the extensions. In order to avoid name clashes, by convention
+      # toolkeys must start by the name of the packages that provides them.
```
Comment (dnjohnstone) [inline doc/dev_guide/writing_extensions.rst]: this was better before.

File: hyperspy/_components/heaviside.py [hunk 37]
```diff
@@ -30,48 +31,42 @@ class HeavisideStep(Component):
...
 
     Parameters
     -----------
     n : float
-
+        Location parameter defining the x position of the step.
     A : float
-
-
+        Height parameter for x>n.
```
Comment (dnjohnstone) [inline hyperspy/_components/heaviside.py]: Should be >=n I think

**Review by dnjohnstone — CHANGES_REQUESTED** [review]
Thanks @jlaehne 

The component written as an expression looks fine except for a minor docstring typo.

I suppose the documentation changes might be a conflict with something that's been merged since this PR was made, I reviewed them before thinking about that....

File: hyperspy/_components/heaviside.py (hunk 37 above)
Comment (jlaehne) [reply hyperspy/_components/heaviside.py]: yep

**Comment by jlaehne** [comment]
Thanks for the review @dnjohnstone and sorry for the confusion with the reverted changes in the documentation. My fault. It was basically reverting changes that have since been merged into RNM. I sorted that out now.

**Comment by dnjohnstone** [comment]
Thanks @jlaehne - I think this PR is now fine. The failing test is, I think, unrelated so I would merge this.

**Comment by jlaehne** [comment]
Test for the function was implemented, so not sure why Codecov fails?

**Comment by tjof2** [comment]
Codecov seems broken more generally

**Review by tjof2 — APPROVED** [review]

**Pull request merged**